src/adhoc.erl

```diff
@@ -105,33 +105,55 @@ produce_response(#adhoc_response{lang = _Lang,
                                  actions = Actions,
                                  notes = Notes,
                                  elements = Elements}) ->
-    SessionID = if is_binary(ProvidedSessionID), ProvidedSessionID /= <<"">> ->
-                        ProvidedSessionID;
-                   true ->
-                        USec = os:system_time(microsecond),
-                        TS = calendar:system_time_to_rfc3339(USec, [{offset, "Z"},
-                                                                    {unit, microsecond}]),
-                        list_to_binary(TS)
-                end,
-    ActionsEls = case Actions of
-                     [] ->
-                         [];
-                     _ ->
-                         ActionsElAttrs = case DefaultAction of
-                                              <<"">> -> [];
-                                              _ -> [{<<"execute">>, DefaultAction}]
-                                          end,
-                         [#xmlel{name = <<"actions">>, attrs = ActionsElAttrs,
-                                 children = [#xmlel{name = Action} || Action <- Actions]}]
-                 end,
-    NotesEls = lists:map(fun({Type, Text}) ->
-                                 #xmlel{name = <<"note">>,
-                                        attrs = [{<<"type">>, Type}],
-                                        children = [#xmlcdata{content = Text}]}
-                         end, Notes),
+    SessionID = ensure_correct_session_id(ProvidedSessionID),
+    ActionsEls = maybe_actions_element(Actions, DefaultAction),
+    NotesEls = lists:map(fun note_to_xmlel/1, Notes),
     #xmlel{name = <<"command">>,
            attrs = [{<<"xmlns">>, ?NS_COMMANDS},
                     {<<"sessionid">>, SessionID},
                     {<<"node">>, Node},
                     {<<"status">>, list_to_binary(atom_to_list(Status))}],
            children = ActionsEls ++ NotesEls ++ Elements}.
+
+-spec ensure_correct_session_id(binary()) -> binary().
+ensure_correct_session_id(SessionID) when is_binary(SessionID), SessionID /= <<"">> ->
+    SessionID;
+ensure_correct_session_id(_) -> 
+    USec = os:system_time(microsecond),
+    TS = calendar:system_time_to_rfc3339(USec, [{offset, "Z"}, {unit, microsecond}]),
+    list_to_binary(TS).
+
+-spec maybe_actions_element([binary()], binary()) -> [exml:element()].
+maybe_actions_element([], _DefaultAction) ->
+    [];
+maybe_actions_element(Actions, <<>>) ->
+    % If the "execute" attribute is absent, it defaults to "next".
+    AllActions = ensure_default_action_present(Actions, <<"next">>),
```

Comment:
Why do you think we need to add "next" here? For me the comment form line 130 means that the attribute can be missing, and then the client will interpret it as "next".

Reply:
This quote from XEP describes it "... it is invalid to return a command that has no "execute" attribute, an actions element, and no <next/> action". When the "execute" attribute is missing client assumes that "next" is the default action therefore it should be in the list of available actions.